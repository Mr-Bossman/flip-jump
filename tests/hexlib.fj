startup

// prints
hex.print 16, vars, 0
output '\n'
hex.print 16, vars, 1
output '\n'
output '\n'


// xor
rep (16, val) xor_all val
output '\n'

// zero
hex.zero  16, vars
hex.print 16, vars, 0
output '\n'
output '\n'

// xor (as vec) once again
hex.xor 16, vars, consts
hex.print 16, vars, 0
output '\n'
output '\n'

// xor_zero
hex.xor_zero 16, vars2, vars
hex.print 16, vars, 0
output '\n'
hex.print 16, vars2, 0
output '\n'
output '\n'

// mov
hex.mov 16, vars, vars2
hex.print 16, vars, 0
output '\n'
hex.print 16, vars2, 0
output '\n'
output '\n'

// set
rep(16, i) hex.set vars+i*dw, 15-i
hex.print 16, vars, 0
output '\n'
hex.print 16, vars2, 0
output '\n'
output '\n'

// swap
hex.swap 16, vars, vars2
hex.print 16, vars, 0
output '\n'
hex.print 16, vars2, 0
output '\n'
output '\n'

// inc1
rep(16, i) inc1_print_carry vars+i*dw, vars2+i*dw
hex.print 16, vars2, 0
output '\n'
hex.print 16, vars, 0
output '\n'
output '\n'

// inc, not
hex.zero 16, vars
hex.inc 16, vars
hex.print 16, vars, 0
output '\n'
hex.set vars, 0xf
hex.inc 16, vars
hex.print 16, vars, 0
output '\n'
hex.zero 16, vars
hex.not 16, vars
hex.print 16, vars, 0
output '\n'
hex.inc 16, vars
hex.print 16, vars, 0
output '\n'
output '\n'

// dec1
hex.mov 16, vars, consts
rep(16, i) dec1_print_borrow vars+i*dw, vars2+i*dw
hex.print 16, vars2, 0
output '\n'
hex.print 16, vars, 0
output '\n'
output '\n'

// dec
hex.zero 16, vars
hex.set vars, 2
hex.dec 16, vars
hex.print 16, vars, 0
output '\n'
hex.dec 16, vars
hex.print 16, vars, 0
output '\n'
hex.dec 16, vars
hex.print 16, vars, 0
output '\n'
hex.dec 16, vars
hex.print 16, vars, 0
output '\n'
output '\n'

// neg
hex.zero 16, vars
hex.neg 16, vars
hex.print 16, vars, 0
output '\n'
hex.set vars+dw, 3
hex.neg 16, vars
hex.print 16, vars, 0
output '\n'
hex.neg 16, vars
hex.print 16, vars, 0
output '\n'
output '\n'

// if_flags
//                    ops = [(randint(0, 15), randint(0,(2**16)-1)) for _ in range(100)]
// hexlib.out line:   for h, f in ops: print((f>>h)&1, end='')
// IF_100_NUMS = hex(sum(h+f*16 << (i*20) for i, (h,f) in enumerate(ops)))
IF_FLAGS_NUMS = 0x2098593d070776a83803ed025fcf2630b539677e0a51278ec3b572bfa969f37eb32fa5a27d38f4f1ba10d5900345ee5d4b50f0d9b73a3ef84d8cbd2258f5a9cdedd958464321a1be0a3d2ba53a12ff684678ccaeea178472da56b4559e3b6e1fb5afd9afc3d69a167c53a95b2516817ba4f9890ded69f06b8e4eae281a4968d57d836496ceea29b290de0a64f6352588dc27b156aa22acc8147ff6400f5d71cd956b7021d2175e2f44c5f78c7d2afc97dacc10e9f17d11206ab17f05ffaf20a7af40b9c09c305a72f0dd8782762922df18191171efd38c65b187350bbecbd27c62c480459289ce5599ceda69a776420065898f22b7cf640a898
rep(100, i) print_if_flags (IF_FLAGS_NUMS>>(i*20))&((i<<20)-1)
output '\n'
output '\n'

// if
def print_if hex @ l0, l1 < IO {
    hex.if hex, l0, l1
    l0: IO+0;l1+dw
    l1: output '1'
}
def print_if0 hex @ l0 < IO {
    hex.if0 hex, l0
    IO+1;l0+dw
    l0: output '0'
}
def print_if1 hex @ l1 < IO {
    hex.if1 hex, l1
    IO+0;l1+dw
    l1: output '1'
}
hex.zero vars
print_if vars
print_if0 vars
print_if1 vars
hex.set vars, 1
print_if vars
print_if0 vars
print_if1 vars
hex.set vars, 8
print_if vars
print_if0 vars
print_if1 vars
hex.set vars, 15
print_if vars
print_if0 vars
print_if1 vars
output '\n'

// if n
def print_if_16 hex @ l0, l1 < IO {
    hex.if 16, hex, l0, l1
    l0: IO+0;l1+dw
    l1: output '1'
}
def print_if0_16 hex @ l0 < IO {
    hex.if0 16, hex, l0
    IO+1;l0+dw
    l0: output '0'
}
def print_if1_16 hex @ l1 < IO {
    hex.if1 16, hex, l1
    IO+0;l1+dw
    l1: output '1'
}
hex.zero 16, vars
print_if_16 vars
print_if0_16 vars
print_if1_16 vars
hex.set vars, 1
print_if_16 vars
print_if0_16 vars
print_if1_16 vars
hex.set vars, 0
hex.set vars+7*dw, 4
print_if_16 vars
print_if0_16 vars
print_if1_16 vars
hex.zero 16, vars
hex.not 16, vars
print_if_16 vars
print_if0_16 vars
print_if1_16 vars
output '\n'
output '\n'


// input4
rep(16, i) input8 vars+i*dw
hex.print 16, vars, 0
input8 garbage
output '\n'

// input_hex
rep(64, i) input_hex_print_error
input8 garbage
output '\n'
output '\n'


loop


vars:   rep(16, j) hex j
vars2:  rep(16, j) hex 0
consts: rep(16, j) hex j
garbage: hex

def xor_all val < vars, consts {
    rep (16, i) hex.xor vars+i*dw, consts+val*dw
    hex.print 16, vars, 0
    output '\n'
}

def inc1_print_carry hex, carry @ output1, end {
    hex.zero carry
    hex.inc1 hex, end, output1
  output1:
    carry+dbit;
  end:
}

def dec1_print_borrow hex, borrow @ output1, end {
    hex.zero borrow
    hex.dec1 hex, end, output1
  output1:
    borrow+dbit;
  end:
}

def print_if_flags hex_flags_val @ l0, l1, hex < IO {
    hex.if_flags hex, hex_flags_val>>4, l0, l1
  hex: hex hex_flags_val & 0xf
  l0:
    IO+0;l1+dw
  l1:
    output '1'
}

def input8 hex < garbage {
    hex.input_ascii hex, garbage
}

def input_hex_print_error @ hex, error, end {
    hex.input_hex hex, error
    hex.print hex, 0

  hex: hex
  error:
    output 'X'
  end:
}
