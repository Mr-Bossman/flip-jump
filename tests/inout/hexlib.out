fedcba9876543210
FEDCBA9876543210

fedcba9876543210
efcdab8967452301
cdef89ab45670123
fedcba9876543210
ba98fedc32107654
efcdab8967452301
89abcdef01234567
fedcba9876543210
76543210fedcba98
efcdab8967452301
45670123cdef89ab
fedcba9876543210
32107654ba98fedc
efcdab8967452301
0123456789abcdef
fedcba9876543210

0000000000000000

fedcba9876543210

0000000000000000
fedcba9876543210

fedcba9876543210
fedcba9876543210

0123456789abcdef
fedcba9876543210

fedcba9876543210
0123456789abcdef

1000000000000000
0fedcba987654321

0000000000000001
0000000000000010
ffffffffffffffff
0000000000000000

0000000000000001
edcba9876543210f

0000000000000001
0000000000000000
ffffffffffffffff
fffffffffffffffe

0000000000000000
ffffffffffffffd0
0000000000000030

0111010110111111001101101100100100111000011101011000100100001100011010100100111100010000010000100110

000111111111
000111111111

fedcba9876543210
0123456789XXXXXXXABCDEFXXXXXXXXXXXXXXXXXXXXXXXXXXabcdefXXXXXXXXX

